lib/improver/nbhood.py

```diff
@@ -246,27 +246,102 @@ def _apply_kernel_for_smoothing(self, cube, ranges):
         # contained within the desired radius.
         kernel = np.ones([int(1 + x * 2) for x in fullranges])
         # Create an open multi-dimensional meshgrid.
-        open_grid = np.ogrid[tuple([slice(-x, x+1) for x in ranges])]
-        if self.unweighted_mode:
+        open_grid = np.array(np.ogrid[tuple([slice(-x, x+1) for x in ranges])])
+        if unweighted_mode:
             mask = np.reshape(
-                np.sum([x ** 2 for x in open_grid]) > np.cumprod(ranges)[-1],
-                np.shape(kernel)
-            )
+                np.sum(open_grid**2) > np.prod(ranges), np.shape(kernel))
         else:
             # Create a kernel, such that the central grid point has the
             # highest weighting, with the weighting decreasing with distance
             # away from the central grid point.
+            open_grid_summed_squared = np.sum(open_grid**2.).astype(float)
             kernel[:] = (
-                (np.cumprod(ranges)[-1] - np.sum([x**2. for x in open_grid])) /
-                np.cumprod(ranges)[-1]
-            )
+                (np.prod(ranges) - open_grid_summed_squared) / np.prod(ranges))
             mask = kernel < 0.
         kernel[mask] = 0.
         # Smooth the data by applying the kernel.
         cube.data = scipy.ndimage.filters.correlate(
             data, kernel, mode='nearest') / np.sum(kernel)
         return cube
 
+
+class NeighbourhoodProcessing(object):
+    """
+    Apply a neigbourhood processing kernel to a thresholded cube.
+
+    When applied to a thresholded probabilistic cube, it acts like a
+    low-pass filter which reduces noisiness in the probabilities.
+
+    The kernel will presently only work with projections in which the
+    x grid point spacing and y grid point spacing are constant over the
+      entire domain, such as the UK national grid projection
+
+    A maximum kernel radius of 500 grid cells is imposed in order to
+    avoid computational ineffiency and possible memory errors.
+
+    """
+
+    # Max extent of kernel in grid cells.
+    MAX_KERNEL_CELL_RADIUS = 500
```

Comment:
This static value and the docstring comments about kernels above should be in the Kernels class.